### ISSUE TITLE
Understanding keras output

### ISSUE THREAD

**Issue by keelinm** [bstriner/keras-adversarial#25]
I am using keras-adversarial for semantic segmentation and I don't understand why loss values output by keras are changing on a model that should not be training (based on the schedule set).  

I have specifically set the schedule so that the discriminator is not training for the first 500 batches, but even so, the discriminator loss values are changing in the keras output from the outset.  Why is this the case?

generator and discriminator models are created and this is the code that follows:

```python
seg = generator(generator.inputs)  # the segmentation of the input images
y_fake = discriminator(generator(generator.inputs)) # output on fake segmentations
y_real = discriminator(discriminator.inputs) # output on real segmentations

# make gan model          
gan_model = Model(generator.inputs + discriminator.inputs, fix_names([tissue_pred, y_fake, y_real], ['tissue_pred', 'y_fake', 'y_real']))
  
gan_model.summary()

# only update the discriminator every 500 batches  
sched = np.ndarray.tolist(np.zeros(500, dtype=np.int8))
sched[499]=1

# adversarial model        
model = AdversarialModel(base_model=gan_model,
                              player_params=[generator.trainable_weights, discriminator.trainable_weights],
                               player_names=["generator", "discriminator"])

# compile  (cross_ent is a measure computing cross entropy loss across channels)
model.adversarial_compile(adversarial_optimizer=AdversarialOptimizerScheduled(schedule=sched),
                                player_optimizers=[Adam(lr=0.001), Adam(lr=0.001)],
                                loss={'seg':cross_ent, 'y_fake':'binary_crossentropy', 'y_real':'binary_crossentropy'})
```

Output is as follows:  I expected that loss values starting with discriminator_ would not change yet, but they do.

Epoch 1/500
53/54 [============================>.] - ETA: 0s - loss: 4.7940 - generator_loss: 2.3954 - generator_tissue_pred_loss: 1.0078 - generator_y_fake_loss: 0.7124 - generator_y_real_loss: 0.6742 - discriminator_loss: 2.3985 - discriminator_tissue_pred_loss: 1.0078 - discriminator_y_fake_loss: 0.6754 - discriminator_y_real_loss: 0.7144 
54/54 [==============================] - 44s - loss: 4.7919 - generator_loss: 2.3945 - generator_tissue_pred_loss: 1.0067 - generator_y_fake_loss: 0.7127 - generator_y_real_loss: 0.6741 - discriminator_loss: 2.3974 - discriminator_tissue_pred_loss: 1.0067 - discriminator_y_fake_loss: 0.6751 - discriminator_y_real_loss: 0.7146 - val_loss: 7.9896 - val_generator_loss: 3.9992 - val_generator_tissue_pred_loss: 2.6059 - val_generator_y_fake_loss: 0.7231 - val_generator_y_real_loss: 0.6692 - val_discriminator_loss: 3.9904 - val_discriminator_tissue_pred_loss: 2.6059 - val_discriminator_y_fake_loss: 0.6641 - val_discriminator_y_real_loss: 0.7195

**Comment by keelinm** [comment]
Actually I just realised the answer myself!  The loss is changing not because the discriminator weights are changing, but only because different data samples are passing through it.    Keras reports the losses for both models even when only one is being trained.

**Comment by keelinm** [comment]
Further to this question, I am not sure I am interpreting my loss output correctly.   As a brief example, how can I account for large differences between generator_y_fake_loss and discriminator_y_fake_loss at every epoch?  As I understood it these are both output from the joint GAN system, just during different training phases, so why should the loss values be so different?
![figure_1](https://user-images.githubusercontent.com/6442576/27225472-3f028f8e-5292-11e7-84ee-928a01ef8420.png)

**Comment by bstriner** [comment]
Those are two different losses based off the same value, `y_fake=D(G(z samples))`. 
* `discriminator_y_fake_loss=mean(-log(1-y_fake))` How well the discriminator categorizes fakes
* `generator_y_fake_loss=mean(-log(y_fake))` How badly the discriminator categorizes fakes / How well the generator confuses the discriminator

Roughly, the relationship is `a=-log(1-e^(-b))` (and vice-versa) but that isn't exact because of the mean. If you did a batch size of one that relationship would work out exactly.

If you are doing WGAN, one will be exactly negative the other. With a GAN, the relationship is more complicated. Roughly, they will start at two different values and tend to converge.

Cheers

**Comment by bstriner** [comment]
If you set batch size to 1, batches per epoch to 1, remove all regularization, and set the optimizer to simultaneous, you should see `discriminator_y_fake_loss=-log(1-e^(-generator_y_fake_loss))`. Values will be slightly different with different settings.
